Move armv8 platforms to be platmods

r1mikey opened the pull request:
With all the work we've done recently to clean up interfaces and platform dependencies we're in a position to remove the `rpi4` and `virt` kernels, leaving the platform-specific code as platmods.

Resurrect `platmod` from the `sun4v` and `sun4v/ontario` directories, integrating them into `armv8` and shuffling the platform code around to simplify the tree and match the way `ontario` was done.

There's loads more exposition in the top commit in this PR ([armv8: Move supported boards to be platmods](https://github.com/richlowe/illumos-gate/pull/126/commits/dbf9e39b887821f84f3da5c7140052fa3e575b6d)).

mail_msg: https://gist.github.com/r1mikey/cf5760e37cc7de3eba4009aba1c6d9f8

Testing
* qemu-virt: https://gist.github.com/r1mikey/86e0ff01f62a1c5034c11af44888d911
* rpi4: https://gist.github.com/r1mikey/7daf92b17acfff7472737487c707ae7d

r1mikey committed:
virt: Move todpl031 to armv8 and rename

The `todpl031` module is not virt-specific - it represents real
hardware, described in DDI0224 Arm PrimeCell Real Time Clock (PL031).
The module is also not named the way we'd prefer (spelling out the
trailing digit).

Move this driver to `armv8` and rename it to `pl03one`. Hook it up to
the build in the new location using a simplified makefile with no
warning gags.

Ideally this driver would attach properly and use DDI for resource
management, but this is particularly hard given how early the TOD
is needed, so I've left this for another time.

r1mikey committed:
rpi4: Rename files and modules to be less generic

The rpi4 code uses very generic names (genet, as an example) for
headers and modules. Make these less generic to avoid accidental
clashes in the future. The new module names use underscores (most
relevant for network drivers where hyphens are not allowed) and
spell out trailing numbers.

While here, clean up `sys/gpio.h` and remove it from `virt`, where
it is unused.

r1mikey committed:
virt: Prepare for platmods

Restructure the `virt` directory to look very similar to
`sun4v/ontario`, which is the reference I'm using for platmods.

There is only one change here:
* What will become the `platmod` (`virt.c`) moves to `os/`.

r1mikey commented:
@richlowe @hadfl I'm only back home late next week, and still collecting a final, clean mail.msg on this change. I'd love to get an early sense of how this is doing on `rpi4` (I don't expect problems, but I can't test for that until late next week).

You'd expect to see something similar to this at boot (changed `unix` path and new `platmod` line):
```
module /platform/armv8/kernel/aarch64/unix: text at [0xfffffffffe000000, 0xfffffffffe116f2f] data at 0xfffffffffe117000
module /kernel/aarch64/genunix: text at [0xfffffffffe1698c0, 0xfffffffffe432dc7] data at 0xfffffffffe432e00
module /platform/RaspberryPi,4/kernel/misc/aarch64/platmod: text at [0xfffffffffe49ff78, 0xfffffffffe4a00df] data at 0xfffffffffe4a00e0
```

Everything else should be unchanged.

I'd also love to get early feedback on the approach I've taken here. It is all based on an existing platform, but others will likely have a deeper understanding of how this is all supposed to work than I do.

hadfl commented:
This boots up fine on the rpi4. But then suddenly panics in a not very helpful way:
```
braich console login:
panic[cpu1]/thread=ffff80004fe6bc20: BAD TRAP: type=25 (#NV2_DATA_ABORT Data abort) rp=ffff80004fe6b940 addr=0 esr=96000004

sched: #NV2_DATA_ABORT Data abort
data abort esr=96000004: Translation fault (level 0) addr=0x0 direction=read
pid=0, pc=0xfffffffffa0ec26c, sp=0xffff80004fe6ba80, spsr=0x60000145
         x0: ffff8003234e6ca8  x1: ffff80004fe6bc20  x2:                0
         x3: ffff80004fe6bc20  x4:                0  x5: ffff80004fe6bc20
         x6: ffff800311a74180  x7:                0  x8: fffffffffe3e4890
         x9: fffffffffe12cf30 x10: fffffffffa0ec250 x11: fffffffffa5d019c
        x12: fffffffffe39b178 x13: fffffffffe39c3a4 x14: fffffffffe39c634
        x15:      132dc8ed9f6 x16: fffffffffe1c8f20 x17:                0
        x18: f8a70e016b19250a x19:                0 x20:                0
        x21: fffffffffa5d0158 x22: fffffffffe39c000 x23:                1
        x24:                0 x25: fffffffffe39c000 x26: ffff800329cfb800
        x27:                0 x28:                0 x29: ffff80004fe6ba80
        x30: fffffffffe097690
         sp: ffff80004fe6ba80  pc: fffffffffa0ec26c spsr:         60000145


dumping to /dev/zvol/dsk/armpool/dump, offset 65536, content: kernel
  0% done: 0 pages dumped, dump failed: error 5

dump failed
```

r1mikey commented:
Thanks for the test!

Can you share the early boot log, where the load addresses for the primary module and dependencies are printed by the kobj code?

hadfl commented:
sure:
```
module /platform/armv8/kernel/aarch64/unix: text at [0xfffffffffe000000, 0xfffffffffe116f2f] data at 0xfffffffffe117000
module /kernel/aarch64/genunix: text at [0xfffffffffe1698c0, 0xfffffffffe432dc7] data at 0xfffffffffe432e00
module /platform/RaspberryPi,4/kernel/misc/aarch64/platmod: text at [0xfffffffffe49ff80, 0xfffffffffe4a0be7] data at 0xfffffffffe4a0be8
illumos Version r1mikey/one-kernel-0-g718a8ed4f9 64-bit
```

Let me know it you need more info, or want the full log.

r1mikey commented:
> sure:
> 
> ```
> module /platform/armv8/kernel/aarch64/unix: text at [0xfffffffffe000000, 0xfffffffffe116f2f] data at 0xfffffffffe117000
> module /kernel/aarch64/genunix: text at [0xfffffffffe1698c0, 0xfffffffffe432dc7] data at 0xfffffffffe432e00
> module /platform/RaspberryPi,4/kernel/misc/aarch64/platmod: text at [0xfffffffffe49ff80, 0xfffffffffe4a0be7] data at 0xfffffffffe4a0be8
> illumos Version r1mikey/one-kernel-0-g718a8ed4f9 64-bit
> ```
> 
> Let me know it you need more info, or want the full log.

That's perfect, thanks. I'll need to dig into the when I'm back in the UK next week. It seems as though a data symbol is not being relocated, which would be `platform_module_list` - I'll need to dig around to figure out what;s going on.

richlowe commented on usr/src/uts/armv8/rpi4/os/rpi4.c:
```diff
@@ -61,8 +62,15 @@
+ * Platform power management drivers list - empty by default
+ */
+char *platform_module_list[] = {
+	(char *)0
```
`NULL` preferably

richlowe commented on usr/src/uts/armv8/virt/Makefile.files:
```diff
@@ -18,17 +18,25 @@
+# Copyright 2006 Sun Microsystems, Inc.  All rights reserved.
+# Use is subject to license terms.
+#
+#ident	"%Z%%M%	%I%	%E% SMI"
```
`git nits` is going to tell you not to add these SCCS `#ident` lines.

richlowe commented on usr/src/uts/armv8/virt/os/virt.c:
```diff
@@ -24,12 +24,25 @@
+ * Platform power management drivers list - empty by default
+ */
+char *platform_module_list[] = {
+	(char *)0
```
`NULL` again, preferably

richlowe commented on usr/src/uts/armv8/rpi4/bcm2711-emmc2/Makefile:
```diff
@@ -34,7 +34,7 @@ UTSBASE	= ../../..
 OBJECTS		= $(BCM2711_EMMC2_OBJS:%=$(OBJS_DIR)/%)
 ROOTMODULE	= $(ROOT_RPI4_DRV_DIR)/$(MODULE)
-CONF_SRCDIR	= $(UTSBASE)/armv8/rpi4/io/$(MODULE)/
+CONF_SRCDIR	= $(UTSBASE)/armv8/rpi4/io
```
We prefer the conf file and the driver be in the same directory.  Despite apparently sun4v not having done that :)

r1mikey replied:
It's probably not obvious in the diffs, but the drivers are now a single C file each and have moved into `armv8/rpi4/io` directly:
```
michael@omnios:/build/arm64-gate/illumos-gate/usr/src/uts$ ls -1 armv8/rpi4/io
bcm2711-emmc2.c
bcm2711-emmc2.conf
bcm2711-genet.c
bcm2711-genet.conf
michael@omnios:/build/arm64-gate/illumos-gate/usr/src/uts$
```

Does this look OK?

richlowe replied:
I've been trying to do one-directory-per-driver (because I have big illumos build changes that depend upon it), but yes this is fine right now

r1mikey replied:
> I've been trying to do one-directory-per-driver (because I have big illumos build changes that depend upon it), but yes this is fine right now

I can change this back easily enough - I'm having to rename bcm2711-genet to bcm2711_genet anyway, so I'm renaming bcm2711-emmc2 to bcm2711_emmctwo at the same time. I'll update to go with one per directory.

richlowe commented on usr/src/uts/armv8/rpi4/bcm2711-genet/Makefile:
```diff
@@ -34,7 +34,7 @@ UTSBASE	= ../../..
 OBJECTS		= $(BCM2711_GENET_OBJS:%=$(OBJS_DIR)/%)
 ROOTMODULE	= $(ROOT_RPI4_DRV_DIR)/$(MODULE)
-CONF_SRCDIR	= $(UTSBASE)/armv8/rpi4/io/$(MODULE)/
+CONF_SRCDIR	= $(UTSBASE)/armv8/rpi4/io
```
Same comment about conf file

richlowe reviewed (commented):
I was trying to avoid using the real name of qemu, just because it's so damn confusing.  Oh well.

It's a shame there's no real way to name pl031 that fits -- it gets away with it somewhat because it has no instances, but still.  I've been only loosely considering naming things, and assuming we'll go through and make it all better somehow later.

I'm so glad we can get back to platmods, at least for FDT.

r1mikey commented:
> I was trying to avoid using the real name of qemu, just because it's so damn confusing. Oh well.

I'm in two minds about this, but I also prefer the "vanity" names - I'll write up something a little more generic than what we have today that can override based on the root FDT node's compatible property when an override is present.

hadfl commented on usr/src/uts/armv8/os/platmod.c:
```diff
@@ -0,0 +1,35 @@
+ * Platform power management drivers list - empty by default
+ */
+char *platform_module_list[] = {
+	(char *)0
```
This should probably be `NULL`, too.

hadfl commented on usr/src/uts/armv8/Makefile.armv8:
```diff
@@ -42,11 +48,11 @@ UNAME_M		= $(PLATFORM)
-# code specific to a given implementation resides in the appropriately
-# named subdirectory.   This requires these platforms to have their
-# own Makefiles to define ROOT_PLAT_DIRS, USR_PLAT_DIRS, etc.
+# IMPLEMENTATIONS is used to designate sun4v machines which have
```
Probably `s/sun4v/armv8`

hadfl reviewed (commented):
I had a look at this and like where it's heading to. Just found a few nits.

r1mikey commented:
> > I was trying to avoid using the real name of qemu, just because it's so damn confusing. Oh well.
> 
> I'm in two minds about this, but I also prefer the "vanity" names - I'll write up something a little more generic than what we have today that can override based on the root FDT node's compatible property when an override is present.

This is done in the latest code.

r1mikey commented:
This is a totally unexpected crash! Here's an analysis of the crashing function:
```
0000000000007250 <dld_fini_ops>:
    7250:       a9bc7bfd        stp     x29, x30, [sp, #-64]!
    7254:       910003fd        mov     x29, sp
    7258:       a9014fe0        stp     x0, x19, [sp, #16]
    725c:       a90257f4        stp     x20, x21, [sp, #32]
    7260:       f9001bf6        str     x22, [sp, #48]
    7264:       f9401c00        ldr     x0, [x0, #56]
        stream = ops->devo_cb_ops->cb_str;
                 ^^^^^^^^^^^^^^^^
    7268:       f9403414        ldr     x20, [x0, #104]
        stream = ops->devo_cb_ops->cb_str;
                      ^^^^^^^^^^^^^^^^^^^ (stream is x20)
    726c:       a9405a95        ldp     x21, x22, [x20]  <<< crash is here, so stream is NULL
        rq = stream->st_rdinit;
        wq = stream->st_wrinit;
        ^^^^^^^^^^^^^^^^^^^^^^^ (both of these lines, rq is x21, wq is x22)
    7270:       f94016b3        ldr     x19, [x21, #40]
        modinfo = rq->qi_minfo;
        ^^^^^^^^^^^^^^^^^^^^^^^ (modinfo is x19)
    7274:       f94016c0        ldr     x0, [x22, #40]
    7278:       eb13001f        cmp     x0, x19
    727c:       540000e0        b.eq    7298 <dld_fini_ops+0x48>  // b.none--
```

I found this by grabbing modinfo before the crash. I then used the ELR (the IP at the time of the crash) to locate the crashing module, then subtracted the crashing address from the module load base. Note that the FAR (fault address register) is 0x0, so this is a NULL dereference. The address I found corresponds to 0x726c in the module object file. Using `objdump -D` gives me the crash location, and that's fairly easy to map to the source code.

So what's clear here is that `stream` is NULL (therefore `ops->devo_cb_ops->cb_str` is also NULL), and it's later dereferenced to get the `rq` and `wq`, which is the faulting `ldp` instruction.

Looking at the register dump we have:
```
sched: #NV2_DATA_ABORT Data abort
data abort esr=96000004: Translation fault (level 0) addr=0x0 direction=read
pid=0, pc=0xfffffffffa0ec26c, sp=0xffff80004847aa80, spsr=0x60000345
         x0: ffff800211b4aca8  x1: ffff80004847ac20  x2:                0
         x3: ffff80004847ac20  x4:                0  x5: ffff80004847ac20
         x6: ffff800209f4c6c0  x7:                0  x8: fffffffffe3e4890
         x9: fffffffffe12cf30 x10: fffffffffa0ec250 x11: fffffffffa5d019c
        x12: fffffffffe39b178 x13: fffffffffe39c3a4 x14: fffffffffe39c634
        x15: 560f1a0600000000 x16: fffffffffe1c8f20 x17:                0
        x18:               15 x19:                0 x20:                0
        x21: fffffffffa5d0158 x22: fffffffffe39c000 x23:                1
        x24:                0 x25: fffffffffe39c000 x26: ffff800216cca800
        x27:                0 x28:                0 x29: ffff80004847aa80
        x30: fffffffffe097690
         sp: ffff80004847aa80  pc: fffffffffa0ec26c spsr:         60000345
```

You'll see that x20 is indeed 0, which is the `steam` variable based on the analysis above.

r1mikey commented:
Confirmed that this patch works around the issue, now the real question to answer is "why is it happening at all"? I'll dig a little more.

```
diff --git a/usr/src/uts/common/io/dld/dld_str.c b/usr/src/uts/common/io/dld/dld_str.c
index 419432ac7d..9a5f934eac 100644
--- a/usr/src/uts/common/io/dld/dld_str.c
+++ b/usr/src/uts/common/io/dld/dld_str.c
@@ -534,6 +534,9 @@ dld_fini_ops(struct dev_ops *ops)
        struct module_info *modinfo;

        stream = ops->devo_cb_ops->cb_str;
+       if (stream == NULL)
+               return;
+
        rq = stream->st_rdinit;
        wq = stream->st_wrinit;
        modinfo = rq->qi_minfo;
```

r1mikey commented:
The underlying reason this is blowing up is that the XXX call in `bcm2711-genet` is failing because `find_mbind` (call chain, from `bcm2711-genet`: `_init` ->`mac_init_ops` -> `ddi_name_to_major` -> `mod_name_to_major` -> `find_mbind `) is failing to find the major for bcm2711-genet, returning DDI_MAJOR_T_NONE. This results in the driver not being identified as GLDv3, which means that the call to `mac_register` in the genet attach function fails and things fall apart.

No, it shouldn't be crashing, but it's also just plain broken right now, likely by the renaming change (though I'm not sure how).

hadfl commented:
It could be because the driver name in `mac_init_ops` does no longer match:
https://github.com/richlowe/illumos-gate/blob/ca831ec99c10e93d886dca5174a47c5ac77e4120/usr/src/uts/armv8/rpi4/io/bcm2711-genet.c#L1365

r1mikey commented:
> It could be because the driver name in `mac_init_ops` does no longer match:
> 
> https://github.com/richlowe/illumos-gate/blob/ca831ec99c10e93d886dca5174a47c5ac77e4120/usr/src/uts/armv8/rpi4/io/bcm2711-genet.c#L1365

Yes, that's it! I was going the (very) long way around to the same answer!

hadfl commented:
> > It could be because the driver name in `mac_init_ops` does no longer match:
> > https://github.com/richlowe/illumos-gate/blob/ca831ec99c10e93d886dca5174a47c5ac77e4120/usr/src/uts/armv8/rpi4/io/bcm2711-genet.c#L1365
> 
> Yes, that's it! I was going the (very) long way around to the same answer!

I just came to the answer because of your excellent debugging and figuring out it happens in `mac_register`. There was no shortcut.

r1mikey commented:
Next minor hurdle: https://github.com/richlowe/illumos-gate/blob/arm64-gate/usr/src/uts/common/os/sunddi.c#L5577

I'll rename the drivers to use underscores instead of hyphens.

r1mikey committed:
rpi4: Prepare for platmods

Restructure the `rpi4` directory to look very similar to
`sun4v/ontario`, which is the reference I'm using for platmods.

There are two meaningful changes here:
* What will become the `platmod` (`rpi4.c`) moves to `os/`.
* Drivers are simplified (the headers are unnecessary).

r1mikey committed:
armv8: Move supported boards to be platmods

Makefiles have been updated with the `platmod` machinery based on the
`sun4v` implementation and `sun4v/ontario` platform. The
`LINKED_PLATFORMS` code has been removed from the install rules so
that (unused) userland directories are not created. The approach taken
here, for the most part, has been to reapply the `rpi4` and `virt`
changes on top of the makefiles from `uts/sun4v/ontario` and to pull
the relevant build machinery into `uts/armv8` from `uts/sun4v`.

Since we already have a `sys/platform_modules.h` in `uts/aarch64`,
which is used by code (via unimplemented weak interfaces) in `genunix`,
I have elected to keep the `sys/platmod.h` in `armv8` as the place
to define the platform interfaces needed by the port, given these are
`armv8`-specific (and in most cases, `rpi4`-specific). The functions
are now declared as weak and checked using the usual pattern at
the callsites.

The platform directories remain unchanged, but a link has been
added to the name that would be defined by the first compatible
entry on the root node of the passed FDT. This allows us to experiment
with removing the "vanity naming" of the root node and make naming
be driven by firmware tables alone.

With the platmod code extracted (`virt/os/virt.c`, `rpi4/os/rpi4.c`)
the kernel is identical between the two platforms. Given this, move
the kernel to live in `uts/armv8`. The kernel is located via the
existing default path fallback mechanism in `inteboot` - no code
changes were needed here.

The platform modules have been updated to elide unimplemented
functions, implement necessary data structures (`platform_module_list`)
and expected functions that are not checked for presence
(`plat_tod_fault`).

hadfl commented:
This looks good to me and I am looking forward to rebase my work in progress on top of this :)

I don't have a strong opinion but given you change the file names, do you want (for consistency) use underscores instead of hyphens for `bcm2835-vcio.h` and `bcm2835-vcprop.h`, too?

r1mikey commented:
> I don't have a strong opinion but given you change the file names, do you want (for consistency) use underscores instead of hyphens for `bcm2835-vcio.h` and `bcm2835-vcprop.h`, too?

I was considering that - also no strong opinion.

r1mikey commented:
> > I don't have a strong opinion but given you change the file names, do you want (for consistency) use underscores instead of hyphens for `bcm2835-vcio.h` and `bcm2835-vcprop.h`, too?
> 
> I was considering that - also no strong opinion.

This is done now.

richlowe commented:
> I'll rename the drivers to use underscores instead of hyphens.

This check is NIC-driver specific (though you'd never guess that from the name and the comment), so you should only need to fix the genet driver, not the rest.

r1mikey commented:
I renamed both for consistency, but happy to undo the eMMC driver if you'd like.

hadfl commented:
> I renamed both for consistency, but happy to undo the eMMC driver if you'd like.

I think, naming them consistently makes sense (I am using the same pattern for `bcm2711_soctemp(4D)`).

richlowe commented:
I had not noticed that this left draft status.  My apologies.
Is it ready to merge?

r1mikey commented:
Yep, nothing outstanding.
